### DISCUSSION TITLE
roc_sndio: [sox] formats.c: no handler for file extension `mp3'

### DISCUSSION THREAD

**Discussion by dshil** [roc-streaming/roc-toolkit#158] [Q&A] [answered]
I fetched the sox source long ago but today I've decided to rebuild the project from scratch:

```
scons clean
scons -Q  --build-3rdparty=all
```

Before this sox was able to read mp3 files without any problems.

@gavv is it a new problem?

**Comment by dshil** [comment]
@gavv, it is a regression after 2c0591a6. Without `--without-opus` I have zero problem with playing mp3.

Could you clarify what was the problem with MacOS build?

**Comment by gavv** [comment]
SoX does not handle mp3 files by itself, as well as many other formats and audio systems. Instead, it relies on an external library, I think it is mp3lame in this case. SoX supports a lot of input and output formats. When it is built, every format may be either disabled, enabled, or enabled as a dynamically loaded plugin.

When you use libsox from your distribution, it is usually configured to handle many formats as dynamic plugins and they can be installed via individual packages (e.g. libsox-fmt-mp3 in Debian). When you ask Roc to build libsox for you, Roc builds SoX as a static library with minimal dependencies, so that libroc will not depend on additional libraries for mp3, flac, ogg, etc.

In general, `--build-3rdparty=all` means: "build minimal required dependencies as static libraries and link them into libroc so that it will be independent of system libraries as much as possible". IIRC the only exceptions now are libc, libasound, and libpulse.

This approach allows us to keep the build script short and not reinvent a package manager. It also allows to produce a rather portable executable which is very handful when cross compiling, building for an old or exotic environment, in other words when you have some difficulties with building the dependencies. In normal situations, however, you should prefer using system libraries instead of --build-3rdparty.

**Comment by gavv** [comment]
I have no idea why disabling opus also disables mp3 files support. I also have no idea why it was enabled before. We probably want to disable it explicitly if it requires an external library.

I think you should check the ./configure output in `./3rdparty/<arch>/build/sox-<version>/build.log` before and after adding `--without-opus`. You can also check the external dependencies of the resulting libsox using ldd or some other tool.

I have disabled opus support because 1) we don't need it anyway 2) SoX was complaining about some opus headers on MacOS.

**Comment by gavv** [comment]
Related: #100.

**Comment by dshil** [comment, marked as answer]
Well, I used the system SoX as it was advised. The problem has gone. Thanks for
the clarification.
